<a href="https://colab.research.google.com/github/shawn-kg/Projects-in-Machine-Learning-and-AI/blob/main/Homework1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Task 1: Machine Learning Problem
A Machine Learning problem I would like to solve using logistic regression is whether or not there is the presence of heart disease in a patient. Logistic regression is the best choice for solving this problem because the target variable (whether or not heart disease is present) is a binary value, it could either be yes there is heart disease or no there is not heart disease. There could also be certain levels or severity of the heart disease measured as well. 

# Task 2: Exploratory Data Analysis
The public data set was found [here](https://www.kaggle.com/ronitf/heart-disease-uci/version/1) through the Kaggle database. In terms of Exploratory Data Analysis I imported the python pandas library and checked if there were any missing values using the isna function. Fortunately there were no missing data points in this data set. 

In [ ]:
import pandas as pd
import numpy as np
import math

# Make numpy values easier to read.
np.set_printoptions(precision=3, suppress=True)

import tensorflow as tf
from tensorflow.keras import layers
from tensorflow.keras.layers.experimental import preprocessing

data = pd.read_csv('drive/MyDrive/heart.csv')

data.head()

age  sex  cp  trestbps  chol  fbs  ...  exang  oldpeak  slope  ca  thal  target
0   63    1   3       145   233    1  ...      0      2.3      0   0     1       1
1   37    1   2       130   250    0  ...      0      3.5      0   0     2       1
2   41    0   1       130   204    0  ...      0      1.4      2   0     2       1
3   56    1   1       120   236    0  ...      0      0.8      2   0     2       1
4   57    0   0       120   354    0  ...      1      0.6      2   0     2       1

[5 rows x 14 columns]

In [ ]:
data.isna()

age    sex     cp  trestbps   chol  ...  oldpeak  slope     ca   thal  target
0    False  False  False     False  False  ...    False  False  False  False   False
1    False  False  False     False  False  ...    False  False  False  False   False
2    False  False  False     False  False  ...    False  False  False  False   False
3    False  False  False     False  False  ...    False  False  False  False   False
4    False  False  False     False  False  ...    False  False  False  False   False
..     ...    ...    ...       ...    ...  ...      ...    ...    ...    ...     ...
298  False  False  False     False  False  ...    False  False  False  False   False
299  False  False  False     False  False  ...    False  False  False  False   False
300  False  False  False     False  False  ...    False  False  False  False   False
301  False  False  False     False  False  ...    False  False  False  False   False
302  False  False  False     False  False  ...    False  False  False  False   False

[303 rows x 14 columns]

# Task 3: Logistic Regression

In [ ]:
# Task 3: Logistic Regression 
# Vanilla Batch Gradient Descent
def sigmoid(a):
  return 1/(1+np.exp(-a))
heart_features = data.copy();
heart_labels = heart_features.pop("target")

learning_rate = 0.01

#print(heart_features)
#print(heart_labels)

heart_features = np.array(heart_features,dtype=np.float128)

#heart_features = heart_features.transpose();

heart_labels = np.array(heart_labels)

initializer = tf.keras.initializers.glorot_normal(3)
weights = initializer(shape=(13, 1))
weights = np.array(weights)
#print(weights)
bias = 0

#print(heart_features)
real_heart_features = heart_features.copy()
#print(weights)
heart_features = heart_features.transpose()
print(heart_features)
print(real_heart_features)


for i in range(100000):
  A = sigmoid(np.dot(weights.transpose(),heart_features) + bias)
  A = np.array(A,dtype = np.float128)
  cost =-1/303 * np.sum(heart_labels * np.log2(A,None,where=(A!=0)) + (1 - heart_labels) * np.log2(1-A,None,where=((1-A)!=0)))
  dw = np.dot(heart_features,(A-heart_labels).transpose())/303
  db = np.sum(A - heart_labels)/303
  weights = weights - learning_rate*dw
  bias = bias - learning_rate*db

  
print(cost)
A[A>=0.5] = 1
A[A<0.5] = 0
print(A)
print(heart_labels)
#print(heart_labels-A)








[[63. 37. 41. ... 68. 57. 57.]
 [ 1.  1.  0. ...  1.  1.  0.]
 [ 3.  2.  1. ...  0.  0.  1.]
 ...
 [ 0.  0.  2. ...  1.  1.  1.]
 [ 0.  0.  0. ...  2.  1.  1.]
 [ 1.  2.  2. ...  3.  3.  2.]]
[[63.  1.  3. ...  0.  0.  1.]
 [37.  1.  2. ...  0.  0.  2.]
 [41.  0.  1. ...  2.  0.  2.]
 ...
 [68.  1.  0. ...  1.  2.  3.]
 [57.  1.  0. ...  1.  1.  3.]
 [57.  0.  1. ...  1.  1.  2.]]
18.939574262339449216
[[1. 0. 1. 1. 0. 0. 0. 1. 0. 1. 0. 1. 1. 1. 1. 1. 1. 0. 0. 1. 0. 1. 1. 0.
  1. 0. 1. 1. 1. 1. 1. 0. 1. 0. 0. 0. 1. 0. 1. 1. 0. 1. 0. 0. 1. 1. 1. 1.
  1. 1. 1. 0. 0. 1. 1. 0. 1. 1. 1. 0. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1.
  1. 0. 1. 0. 0. 1. 1. 0. 1. 1. 1. 1. 0. 0. 0. 1. 1. 0. 1. 0. 0. 0. 1. 0.
  0. 0. 0. 0. 1. 0. 1. 1. 1. 1. 1. 0. 1. 1. 0. 1. 1. 0. 1. 1. 1. 1. 1. 0.
  0. 1. 1. 1. 1. 1. 0. 1. 1. 0. 1. 1. 1. 1. 1. 1. 1. 1. 0. 0. 1. 0. 1. 0.
  1. 0. 1. 1. 1. 1. 0. 0. 1. 1. 1. 0. 1. 1. 0. 1. 1. 1. 1. 0. 0. 0. 0. 0.
  0. 0. 0. 0. 0. 0. 0. 0. 0. 1. 0. 0. 0. 0. 1. 0. 0. 0. 0. 0. 0. 0. 0. 0.
  

In [ ]:
 # Stochastic Gradient Descent
initializer = tf.keras.initializers.glorot_normal(3)
weights = initializer(shape=(13, 1))
weights = np.array(weights)
 
 
print(weights)
 
 
heart_features = np.array(heart_features)
real_heart_features = heart_features.transpose()
print(heart_features)
data = np.array(data)
 
for i in range(10):
  np.random.shuffle(data)
  for ex in data:
    Y = ex.pop("target")
    print(Y)
    print(ex)
    A = sigmoid(np.dot(weights.transpose(),ex) + bias)
    print(A)
    A = np.array(A,dtype = np.float128)
    #cost =-1/303 * np.sum(heart_labels * np.log2(A,None,where=(A!=0)) + (1 - heart_labels) * np.log2(1-A,None,where=((1-A)!=0)))
    dw = np.dot(ex,(A-heart_labels).transpose())/303
    db = np.sum(A - heart_labels)/303
    weights = weights - learning_rate*dw
    bias = bias - learning_rate*db
 
print(cost)
A[A>=0.5] = 1
A[A<0.5] = 0
print(A)
print(heart_labels)



I tried to implement Stochastic Gradient descent above but I had great trouble understanding it. This is my first Machine Learning and AI class but I hope I can improve my understanding as the class goes on. 

#Task 4: Optimizers

In [ ]:
model = tf.keras.Sequential([layers.Dense(303),layers.Dense(1)])
model.compile(loss = tf.losses.MeanSquaredLogarithmicError(),
                      optimizer = tf.optimizers.Adam())
model.fit(heart_features, heart_labels, epochs=10)

Epoch 1/10
10/10 [==============================] - 0s 3ms/step - loss: 2.1892
Epoch 2/10
10/10 [==============================] - 0s 2ms/step - loss: 0.2616
Epoch 3/10
10/10 [==============================] - 0s 2ms/step - loss: 0.2616
Epoch 4/10
10/10 [==============================] - 0s 2ms/step - loss: 0.2616
Epoch 5/10
10/10 [==============================] - 0s 3ms/step - loss: 0.2616
Epoch 6/10
10/10 [==============================] - 0s 2ms/step - loss: 0.2616
Epoch 7/10
10/10 [==============================] - 0s 2ms/step - loss: 0.2616
Epoch 8/10
10/10 [==============================] - 0s 2ms/step - loss: 0.2616
Epoch 9/10
10/10 [==============================] - 0s 3ms/step - loss: 0.2616
Epoch 10/10
10/10 [==============================] - 0s 2ms/step - loss: 0.2616


In [ ]:
model = tf.keras.Sequential([layers.Dense(303),layers.Dense(1)])
model.compile(loss = tf.losses.MeanSquaredLogarithmicError(),
                      optimizer = tf.optimizers.RMSprop())
model.fit(heart_features, heart_labels, epochs=10)

Epoch 1/10
10/10 [==============================] - 0s 2ms/step - loss: 0.2616
Epoch 2/10
10/10 [==============================] - 0s 2ms/step - loss: 0.2616
Epoch 3/10
10/10 [==============================] - 0s 2ms/step - loss: 0.2616
Epoch 4/10
10/10 [==============================] - 0s 2ms/step - loss: 0.2616
Epoch 5/10
10/10 [==============================] - 0s 2ms/step - loss: 0.2616
Epoch 6/10
10/10 [==============================] - 0s 2ms/step - loss: 0.2616
Epoch 7/10
10/10 [==============================] - 0s 2ms/step - loss: 0.2616
Epoch 8/10
10/10 [==============================] - 0s 2ms/step - loss: 0.2616
Epoch 9/10
10/10 [==============================] - 0s 2ms/step - loss: 0.2616
Epoch 10/10
10/10 [==============================] - 0s 3ms/step - loss: 0.2616


#Conclusion
In conclusion it seems that both Adam and RMSProp provided the same optimization for the data set. Here both optimizers were able to reduce the cost down to 0.2616 in just 10 epochs. Comparing this to the vanilla batch gradient descent is astonishing. It took 100,000 epochs to just get it down to 18 and I could not get it down any further. Optimization should definitely be used for this task at hand because it does not jump around the minimum and slows down its descent as it gets closer and closer to a minimum. This is especially helpful in this data set where heart disease is a danger to human lives. False negatives in these cases cannot be an option. 